### PR TITLE
N-ary ops

### PR DESCRIPTION
**Pull request by alexkeizer**
Generalize Ops to take an arbitrary (but static / non-variadic) number of arguments

### CODE REVIEW

File: SSA/Experimental/ErasedContext.lean [hunk 1]
```diff
@@ -191,26 +191,18 @@ theorem Valuation.snoc_toSnoc {Γ : Ctxt} {t t' : Ty} (s : Γ.Valuation) (x : t.
...
 
 namespace VarSet
 
 /-- A `VarSet` with exactly one variable `v` -/
 @[simp]
 def ofVar {Γ : Ctxt} (v : Γ.Var t) : VarSet Γ :=
-  fun t' => if ty_eq : t = t' then {ty_eq ▸ v} else {}
-
-@[simp]
-instance : EmptyCollection (VarSet Γ) := ⟨fun _ => ∅⟩
```
Comment (goens) [inline SSA/Experimental/ErasedContext.lean:205]: why do we want to remove these instances?

File: SSA/Experimental/IntrinsicAsymptotics.lean [hunk 2]
```diff
@@ -9,83 +9,133 @@ import Mathlib.Tactic.Ring
...
+    ∀ {l : List Ty}, B → HVector A l → B
+  | [], b, .nil => b
+  | t::_, b, .cons a as => foldl f (f t b a) as
+
+@[reducible]
+def Op.denote : (op : Op) →
+    HVector Ty.toType op.sig → op.outTy.toType
   | .cst n, _ => n
-  | .add a b, Γv => Γv a + Γv b
+  | .add, .cons a (.cons b .nil) => a + b
```
Comment (goens) [inline SSA/Experimental/IntrinsicAsymptotics.lean:83]: that's pretty cool that we can just pattern match on these like that! Do you think it makes sense to add syntatic sugar for the pattern matching, so we don't have to call these constructors explicitly and it just sort of "looks like a tuple/list"?

File: SSA/Experimental/IntrinsicAsymptotics.lean [hunk 3]
```diff
@@ -720,38 +901,38 @@ theorem denote_splitProgramAt {pos : ℕ} {prog : ICom Γ₁ t}
 `target`. If it can match the variables, it inserts `rhs` into the program
 with the correct assignment of variables, and then replaces occurences
 of the variable at position `pos` in `target` with the output of `rhs`.  -/
```
Comment (goens) [inline SSA/Experimental/IntrinsicAsymptotics.lean:903]: it seems like the docstring doesn't match the function anymore - could we update the docstring?

File: SSA/Experimental/IntrinsicAsymptotics.lean [hunk 4]
```diff
@@ -773,29 +954,27 @@ structure PeepholeRewrite (Γ : List Ty) (t : Ty) where
...
     rcases v with ⟨i, hi⟩
-    simp only [Erased.out_mk] at hi  
+    simp only [Erased.out_mk] at hi
     rcases List.get?_eq_some.1 hi with ⟨h', rfl⟩
     simp at h'
     convert h ⟨i, h'⟩
   . intro h i
     apply h
 
-def rewritePeepholeAt (pr : PeepholeRewrite Γ t)
```
Comment (goens) [inline SSA/Experimental/IntrinsicAsymptotics.lean:794]: maybe this could also get a short docstring?

**Review by goens — APPROVED** [review]
Left a few minor comments; I'm happy to merge this ofc.

File: SSA/Experimental/ErasedContext.lean (hunk 1 above)
Comment (alexkeizer) [reply SSA/Experimental/ErasedContext.lean:205]: Now that `VarSet` is just an abbrev for a `Finset _`, these instances are found directly as the `Finset` instances

**Pull request merged**